src/libs/relay/conduit_relay_mpi.cpp

```diff
@@ -113,32 +113,235 @@ rank(MPI_Comm mpi_comm)
+
+    switch(dt)
+    {
+        // string type
+        case MPI_CHAR:  res = DataType::CHAR8_STR_ID;  break;
+
+        // signed integer types
+        case MPI_SHORT: res = CONDUIT_NATIVE_SHORT_ID; break;
+        case MPI_INT:   res = CONDUIT_NATIVE_INT_ID;   break;
+        case MPI_LONG:  res = CONDUIT_NATIVE_LONG_ID;  break;
+
+        #if defined(CONDUIT_USE_LONG_LONG)
+        case MPI_LONG_LONG: res = CONDUIT_NATIVE_LONG_LONG_ID; break;
+        #endif
+
+        // unsigned integer types 
+        case MPI_BYTE:           res = CONDUIT_NATIVE_UNSIGNED_CHAR_ID;  break;
+        case MPI_UNSIGNED_CHAR:  res = CONDUIT_NATIVE_UNSIGNED_CHAR_ID;  break;
+        case MPI_UNSIGNED_SHORT: res = CONDUIT_NATIVE_UNSIGNED_SHORT_ID; break;
+        case MPI_UNSIGNED:       res = CONDUIT_NATIVE_UNSIGNED_INT_ID;   break;
+        case MPI_UNSIGNED_LONG:  res = CONDUIT_NATIVE_UNSIGNED_LONG_ID;  break;
+        
+        #if defined(CONDUIT_USE_LONG_LONG)
+        case MPI_UNSIGNED_LONG_LONG:
+        {
+            res = CONDUIT_NATIVE_UNSIGNED_LONG_LONG_ID;
+            break;
+        }
+        #endif
+
+        // floating point types
+
+        case MPI_FLOAT:  res = CONDUIT_NATIVE_FLOAT_ID;  break;
+        case MPI_DOUBLE: res = CONDUIT_NATIVE_DOUBLE_ID; break;
+
+        #if defined(CONDUIT_USE_LONG_DOUBLE)
+        case MPI_LONG_DOUBLE: res = CONDUIT_NATIVE_LONG_DOUBLE_ID; break;
+        #endif
+
+    }
+
+    return res;
+}
+
 
 //---------------------------------------------------------------------------//
 int 
-send(Node &node, int dest, int tag, MPI_Comm comm)
+send_using_schema(const Node &node, int dest, int tag, MPI_Comm comm)
 { 
+    std::string snd_schema ="";
+    const void *snd_ptr = node.contiguous_data_ptr();
+    int         snd_data_size = node.total_bytes_compact();
+
+    Node n_msg;
+    if(snd_ptr != NULL && 
+       node.is_compact() )
+    {
+        snd_schema = node.schema().to_json();
+        n_msg["schema_len"].set((int64)snd_schema.length());
+        n_msg["schema"].set_external_char8_str(const_cast<char*>(snd_schema.c_str()));
+        n_msg["data"].set_external((uint8*)snd_ptr,(index_t)snd_data_size);
+    }
+    else
+    {
+        // order matters for compact result, init schema len and schema to empty
+        // to create proper order
+        n_msg["schema_len"].set(DataType::empty());
+        n_msg["schema"].set(DataType::empty());
+        
+        // compact the data into our msg node
+        node.compact_to(n_msg["data"]);
+        // wire  up the 
+        snd_schema = n_msg["data"].schema().to_json();
+        n_msg["schema_len"].set((int64)snd_schema.length());
+        n_msg["schema"].set_external_char8_str(const_cast<char*>(snd_schema.c_str()));
+    }
+    
 
-    Schema schema_c;
-    node.schema().compact_to(schema_c);
-    std::string schema = schema_c.to_json();
-    int schema_len = schema.length() + 1;
+    Node n_msg_compact;
+    n_msg.compact_to(n_msg_compact);
```

Comment:
Is it possible to start with this compacted schema and then push things into it in the if/else above? It's a pretty simple schema and you would just have to preallocate the data. Otherwise, looks good.